Add assoc_in function.

roryk opened the pull request:
Hi everyone, thanks for making toolz, it is very elegant.

I think `assoc_in` should exist because it is a commonly used idiom when working with nested dictionaries. This doesn't fit the contribution guide of providing a significantly different functionality but it does provide what I think might be expected functionality. There's `update_in` and `get_in` to reach into nested dictionaries, but to do an `assoc_in` you have to figure out to call `update_in` with a noop function and a default value.

It's a wrapper around an existing function so it won't cause much maintenance overhead. Let me know what you think. Thanks again!

mrocklin commented:
Other than the small test failures this looks good to me.  You might also want to add `assoc_in` to `doc/source/api.rst`

roryk commented:
Sweet-- thanks @mrocklin for getting back to me so fast. I added `assoc_in` and fixed the test errors and squashed everything so the history isn't gross.

mrocklin commented:
+1

Merging in 24 hours if no comments

eriknw commented:
+1.  As you said, this is expected functionality for those coming from a functional programming background.

@roryk, would you like to add yourself to AUTHORS.md?

roryk commented:
Thanks, that is very generous. Gave props to myself on there.

roryk commented:
Thanks @mrocklin and @eriknw, you rock.
